cfgrammar/src/lib/yacc/parser.rs

```diff
@@ -326,6 +335,38 @@ impl YaccParser {
         }
     }
 
+    fn parse_action(&mut self, i: usize) -> YaccResult<(usize, String)> {
+        let mut j = i;
+        let mut c = 0; // Count braces
+        while j < self.src.len() {
+            let ch = self.src[j..].chars().next().unwrap();
+            c += match ch {
+                '{' => 1,
+                '}' => -1,
+                _ => 0
+            };
+            if c == 0 {
+                break;
+            }
+            j += ch.len_utf8();
+        }
+        if c > 0 {
+            Err(self.mk_error(YaccParserErrorKind::IncompleteAction, j))
+        } else {
+            let s = self.src[i + 1..j - 1].trim().to_string();
+            Ok((j + 1, s))
+        }
+    }
+
+    fn parse_programs(&mut self, i: usize) -> YaccResult<usize> {
+        let mut j = i + 2; // Skip %%
+        j = self.parse_ws(j)?;
```

Comment:
I don't think you need to introduce `j` here: you can keep overwriting `i`.

Reply:
Don't I have to make `i` mutable then?

Reply:
Yes, but that's transparent to the caller, so it's fine to do so.